### PR TITLE
z3: new, 4.8.9

### PR DESCRIPTION
**Pull request by cthbleachbit**
Topic Description
-----------------

This PR introduces a new package: Z3, an efficient constraint solver and theorem prover utility and its python bindings.

Package(s) Affected
-------------------

`z3` 4.8.9

Architectural Progress
----------------------

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`

Secondary Architectural Progress
--------------------------------

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`

Post-Merge Architectural Progress
---------------------------------

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`

Post-Merge Secondary Architectural Progress
-------------------------------------------

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`

### CODE REVIEW

**Comment by cthbleachbit** [comment]
Testing Procedures
--------------------------

For the main program:

```
cth451@cth-desktop-dorm [ ~ ] $ cat /tmp/test.smt 
(declare-fun x () Int)
(declare-fun y () Int)
(declare-fun z () Int)
(assert (>= (* 2 x) (+ y z)))
(declare-fun f (Int) Int)
(declare-fun g (Int Int) Int)
(assert (< (f x) (g x x)))
(assert (> (f y) (g x x)))
(check-sat)
(get-model)
(push)
(assert (= x y))
(check-sat)
(pop)
(exit)
cth451@cth-desktop-dorm [ ~ ] $ z3 /tmp/test.smt 
sat
(model 
  (define-fun y () Int
    3)
  (define-fun x () Int
    2)
  (define-fun z () Int
    1)
  (define-fun g ((x!0 Int) (x!1 Int)) Int
    0)
  (define-fun f ((x!0 Int)) Int
    (ite (= x!0 3) 1
      (- 1)))
)
unsat
```

For python bindings:
```
cth451@cth-desktop-dorm [ ~ ] $ python3
Python 3.8.6 (default, Oct 21 2020, 14:11:15) 
[GCC 9.3.1 20200616 (AOSC OS, Core)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from z3 import *
>>> x = Int('x')
>>> y = Int('y')
>>> z = Int('z')
>>> f = Function('f', IntSort(), IntSort())
>>> g = Function('g', IntSort(), IntSort(), IntSort())
>>> s = Solver()
>>> s.add(2 * x >= y + z)
>>> s.add(f(x) < g(x,x))
>>> s.add(f(y) > g(x,x))
>>> s.check()
sat
>>> s.model()
[y = 3,
 x = 2,
 z = 1,
 g = [else -> 0],
 f = [3 -> 1, else -> -1]]
>>>
```

File: extra-scientific/z3/autobuild/beyond [hunk 1]
```diff
@@ -0,0 +1,7 @@
+cp -vr ${PKGDIR}/usr/lib/python3.* ${PKGDIR}/usr/lib/python2.7
```
Comment (MingcongBai) [inline extra-scientific/z3/autobuild/beyond]: Dot `abinfo` throughout custom scripts - this is required by the [Styling Manual's Scripting section](https://wiki.aosc.io/developer/packaging/package-styling-manual/#scripting).

File: extra-scientific/z3/autobuild/beyond [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+python2 -m compileall "$PKGDIR"/usr/lib/python2.*/site-packages/z3
```
Comment (MingcongBai) [inline extra-scientific/z3/autobuild/beyond]: Use `$ABPY2VER` and `$ABPY3VER` here.

File: extra-scientific/z3/autobuild/defines [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+PKGDES="Z3 Efficient Theorem Prover"
+PKGSEC=math
+PKGDEP="glibc python-3 python-2"
+BUILDDEP="setuptools gcc-runtime"
+
+CMAKE_AFTER="-DCMAKE_INSTALL_LIBDIR=lib \
+             -DZ3_LINK_TIME_OPTIMIZATION=1 \
+             -DZ3_BUILD_PYTHON_BINDINGS=1 \
+             -DPYTHON_EXECUTABLE=/usr/bin/python3"
+ABTYPE=cmake
```
Comment (MingcongBai) [inline extra-scientific/z3/autobuild/defines]: Does `cmakeninja` not work?

File: extra-scientific/z3/spec [hunk 4]
```diff
@@ -0,0 +1,3 @@
+VER=4.8.9
+SRCS="tarball::https://github.com/Z3Prover/z3/archive/z3-${VER}.tar.gz"
```
Comment (MingcongBai) [inline extra-scientific/z3/spec]: `tbl`, not `tarball`.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Review by MingcongBai — APPROVED** [review]

File: extra-scientific/z3/autobuild/beyond [hunk 5]
```diff
@@ -0,0 +1,9 @@
+abinfo Installing python module for python 2
+cp -vr ${PKGDIR}/usr/lib/python${ABPY3VER} ${PKGDIR}/usr/lib/python${ABPY2VER}
```
Comment (MingcongBai) [inline extra-scientific/z3/autobuild/beyond]: Another issue, `$PKGDIR` should be quoted.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Comment by MingcongBai** [comment]
Another issue is that some CMake options are invalid, specially the part about enabling Python bindings.

```
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_PYTHON_BINDINGS
    LIB_INSTALL_DIR
    SYSCONF_INSTALL_DIR
```

**Comment by Cyanoxygen** [comment]
### Testing in `arm64`
- native z3

image

- python3 binding
image

**Comment by MingcongBai** [comment]
The newest scripts doesn't seem to produce a working build.

```
-- Z3_COMPONENT_EXTRA_INCLUDE_DIRS: /var/cache/acbs/build/acbs.tychw057/z3-z3-4.8.9/build-py2/src;/var/cache/acbs/build/acbs.tychw057/z3-z3-4.8.9/src
-- Z3_DEPENDENT_EXTRA_CXX_LINK_FLAGS: 
-- CMAKE_INSTALL_LIBDIR: "lib              -DZ3_LINK_TIME_OPTIMIZATION=ON              -DZ3_BUILD_PYTHON_BINDINGS=ON              -DPYTHON_EXECUTABLE=/usr/bin/python2"
-- CMAKE_INSTALL_BINDIR: "bin"
-- CMAKE_INSTALL_INCLUDEDIR: "include"
-- CMAKE_INSTALL_PKGCONFIGDIR: "lib              -DZ3_LINK_TIME_OPTIMIZATION=ON              -DZ3_BUILD_PYTHON_BINDINGS=ON              -DPYTHON_EXECUTABLE=/usr/bin/python2/pkgconfig"
-- CMAKE_INSTALL_Z3_CMAKE_PACKAGE_DIR: "lib              -DZ3_LINK_TIME_OPTIMIZATION=ON              -DZ3_BUILD_PYTHON_BINDINGS=ON              -DPYTHON_EXECUTABLE=/usr/bin/python2/cmake/z3"
```

And finally...

```
cp: cannot stat '/var/cache/acbs/build/acbs.tychw057/z3-z3-4.8.9/build-py2/inst/usr/lib/python2.7': No such file or directory
```

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**